src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java

```diff
@@ -0,0 +1,254 @@
+                new MyEvent(7, ZonedDateTime.of(2019, 3, 30, 2, 10, 0, 0, zone)),
+                new MyEvent(1, ZonedDateTime.of(2019, 3, 31, 1, 39, 0, 0, zone)),
+                new MyEvent(2, ZonedDateTime.of(2019, 3, 31, 2, 0, 0, 0, zone)),
+                new MyEvent(7, ZonedDateTime.of(2019, 3, 31, 2, 10, 0, 0, zone)),
+                new MyEvent(7, ZonedDateTime.now())
+        );
+        final List<ExpectedResult> expectedValues = Arrays.asList(
+                new ExpectedResult(1,
+                        10,
+                        ZonedDateTime.of(2019, 3, 29, 18, 0, 0, 0, zone),
+                        ZonedDateTime.of(2019, 3, 30, 18, 0, 0, 0, zone)),
+                new ExpectedResult(1,
+                        10,
+                        ZonedDateTime.of(2019, 3, 30, 18, 0, 0, 0, zone),
+                        ZonedDateTime.of(2019, 3, 31, 18, 0, 0, 0, zone))
+        );
+        kafkaStreamsMagic(inputValues, expectedValues, zone);
+    }
+
+    @Test
+    public void shouldSumNumbersWithTwoWindowsAndDSTTimezone() {
+        //This test illustrate problems with daylight savings
+        //Some timezone have daylight savings time (DST) resulting in two days in year that have either 23 or 25 hours.
+        //Kafka streams currently support only fixed period for the moment.
+        ZoneId zoneWithDST = ZoneId.of("Europe/Paris");
+
+        final List<MyEvent> inputValues = Arrays.asList(
+                new MyEvent(1, ZonedDateTime.of(2019, 3, 30, 1, 39, 0, 0, zoneWithDST)),
+                new MyEvent(2, ZonedDateTime.of(2019, 3, 30, 2, 0, 0, 0, zoneWithDST)),
+                new MyEvent(7, ZonedDateTime.of(2019, 3, 30, 2, 10, 0, 0, zoneWithDST)),
+                new MyEvent(1, ZonedDateTime.of(2019, 3, 31, 1, 39, 0, 0, zoneWithDST)),
+                new MyEvent(2, ZonedDateTime.of(2019, 3, 31, 2, 0, 0, 0, zoneWithDST)),
+                new MyEvent(7, ZonedDateTime.of(2019, 3, 31, 2, 10, 0, 0, zoneWithDST)),
+                new MyEvent(7, ZonedDateTime.now())
+        );
+        final List<ExpectedResult> expectedValues = Arrays.asList(
+                new ExpectedResult(1,
+                        10,
+                        ZonedDateTime.of(2019, 3, 29, 18, 0, 0, 0, zoneWithDST),
+                        ZonedDateTime.of(2019, 3, 30, 18, 0, 0, 0, zoneWithDST)),
+                new ExpectedResult(1,
+                        10,
+                        ZonedDateTime.of(2019, 3, 30, 18, 0, 0, 0, zoneWithDST),
+                        //This get one extra hour due to time shift in Daylight saving
+                        //As a user i would expect it to end on 31th at 6pm.
+
+                        //The limitation seems to come from TimeWindowSerializer / TimeWindowDeserializer as we serialize only start date.
+                        //Suggestion: By serializing both start and end of window, we could support more complex cases with non fixed time window and address daylight savings on daily windows.
```

Comment:
Yes, this is exactly right. Feel free to open a ticket for this, and thanks for providing an example of how to accomplish custom time windows that explicitly includes this edge case!

Reply:
Also, it's worth noting that days will have different lengths under other conditions than just leap year. Leap seconds, for example.

This is actually part of the reason we use "milliseconds since the epoch" as our clock, and align our time windows to some interval on that clock. That way, the internal clock is both monotonic and regular, neither of which are properties that real-life clocks possess.

Of course, this is an academic choice, since it simplifies the implementation at the expense of realistic use cases like "aggregate daily at midnight".